app/views/courses/TeacherClassView.coffee

```diff
@@ -390,35 +386,92 @@ module.exports = class TeacherClassView extends RootView
-    @state.set errors: { assigningToNobody, assigningToUnenrolled }
+    @state.set errors: { assigningToNobody }
     return if assigningToNobody
-    @assignCourse courseID, members
+    @assignCourse courseID, selectedIDs
     window.tracker?.trackEvent 'Teachers Class Students Assign Selected', category: 'Teachers', classroomID: @classroom.id, courseID: courseID, ['Mixpanel']
 
-  # TODO: Move this to the model. Use promises/callbacks?
   assignCourse: (courseID, members) ->
-    courseInstance = @courseInstances.findWhere({ courseID, classroomID: @classroom.id })
-    if courseInstance
-      courseInstance.addMembers members
-    else
-      courseInstance = new CourseInstance {
-        courseID,
-        classroomID: @classroom.id
-        ownerID: @classroom.get('ownerID')
-        aceConfig: {}
-      }
-      @courseInstances.add(courseInstance)
-      courseInstance.save {}, {
-        success: ->
-          courseInstance.addMembers members
-      }
-    null
+    courseInstance = null
+    numberEnrolled = 0
+    remainingSpots = 0
+    
+    return Promise.resolve()
+    .then =>
+      courseInstance = @courseInstances.findWhere({ courseID, classroomID: @classroom.id })
+      if not courseInstance
+        courseInstance = new CourseInstance {
+          courseID,
+          classroomID: @classroom.id
+          ownerID: @classroom.get('ownerID')
+          aceConfig: {}
+        }
+        courseInstance.notyErrors = false # handling manually
+        @courseInstances.add(courseInstance)
+        return courseInstance.save()
+        
+    .then =>
+      availablePrepaids = @prepaids.filter((prepaid) -> prepaid.status() is 'available')
+      unenrolledStudents = _(members)
+        .map((userID) => @students.get(userID))
+        .filter((user) => user.prepaidStatus() isnt 'enrolled')
+        .value()
+      totalSpotsAvailable = _.reduce(prepaid.openSpots() for prepaid in availablePrepaids, (val, total) -> val + total) or 0
+      if totalSpotsAvailable < _.size(unenrolledStudents)
+        modal = new CoursesNotAssignedModal({
+          selected: members.length
+          totalSpotsAvailable
+          unenrolledStudents: _.size(unenrolledStudents)
+        })
+        @openModalView(modal)
+        error = new Error('Not enough licenses available')
+        error.handled = true
+        throw error
+        
+      numberEnrolled = _.size(unenrolledStudents)
+      remainingSpots = totalSpotsAvailable - numberEnrolled
+      
+      requests = []
+      for prepaid in availablePrepaids
+        for i in _.range(prepaid.openSpots())
+          break unless _.size(unenrolledStudents) > 0
+          user = unenrolledStudents.shift()
+          requests.push(prepaid.redeem(user))
+      
+      @trigger 'begin-redeem-for-assign-course' 
+      return $.when(requests...)
+
+    .then =>
+      prepaid.fetch() for prepaid in @prepaids.find() # make sure these are completely up to date, but don't block on them.
```

Comment:
Ah, good catch. I'm firing off a bunch of parallel redeem requests in the previous `then`, and I'd be very lucky if the last one to return includes the changes made by all the others. So I want the view to reload all the prepaids to make sure they contain the proper numbers of redeemers.